Evolutionary_loop/Loop_Parts/Part_B/part_b_job1.py

```diff
@@ -0,0 +1,229 @@
+                   f'>> {sim_path}', shell=True)
+
+        else:                                  # straight-side asymmetric bicone
+            sp.run(f'cat {xmacros_dir}/simulationPECmacroskeleton_Sep.txt'
+                   f'>> {sim_path}', shell=True)
+            sp.run(f'cat {xmacros_dir}/simulationPECmacroskeleton2_Sep.txt'
+                   f'>> {sim_path}', shell=True)
+    else:                                      # curved-side bicone
+        sp.run(f'cat {xmacros_dir}/simulationPECmacroskeleton_curved.txt'
+               f'>> {sim_path}', shell=True)
+        sp.run(f'cat {xmacros_dir}/simulationPECmacroskeleton2_curved.txt'
+               f'>> {sim_path}', shell=True)
+
+
+## SED-TION (Ha! Get it? Like...section?)
+
+    # sed swaps out the hardcoded word "fileDirectory" inside the xmacro.
+    # option "i" of sed: swap "in-place" instead of outputting to terminal.
+    # The empty quotes after sed -i are for MacOS compatibility:
+    # https://stackoverflow.com/questions/12272065/sed-undefined-label-on-macos
+    sp.run(f'sed -i "" "s|fileDirectory|{working_dir}/Generation_Data|" '
+           f'{sim_path}', shell=True)
+
+    # Change gridsize by the same factor used for changing antenna size.
+    xmacro_skeleton_default_gridsize = 0.1
+    grid_size = "%.6f" % (xmacro_skeleton_default_gridsize/geo_factor)
+
+    sp.run(f'sed -i "" "s/var gridSize = 0.1;/var gridSize = {grid_size};/" '
+           f'{sim_path}', shell=True)
+
+
+## XF SECTION
+
+    os.chmod(xmacros_dir, 0o775)
+    # load XFdtd on OSC
+    p0 = sp.run(f'module load xfdtd/7.9.2.2', shell=True, capture_output=True)
+    if p0.returncode: # print out the error message
+        print('\nError at part_b_job1.py: XF SECTION while loading module\n'+
+              p0.stderr.decode())
+
+    # load simulation_PEC.xmacro
+    p1 = sp.run(f'xfdtd {xf_proj} --execute-macro-script={sim_path}', 
+                shell=True, capture_output=True) 
+    if p1.returncode: # print out the error message
+        print('\nError at part_b_job1.py XF SECTION while loading xmacro\n'+
+              p1.stderr.decode())
+
+    # Determine the number of simultaneously running jobs in the SLURM job-array
+    if npop <= num_keys:
+        batch_size = npop       # quite unlikely, since we only have 5 keys
+    else:
+        batch_size = num_keys
+
+
+## SIMULATION JOB SUBMISSION
+    # Submit a job-array to OSC; each job in the array is responsible for
+    # simulating one antenna in the current generation. Make the SLURM job-name
+    # the same as run_name so that we can use SLURM variables.
+    p2 = sp.run(f'sbatch --array=1{npop}%{batch_size} '
+                f'--job-name={run_name} '
+                f'--export=ALL,WorkingDir={working_dir},RunName={run_name}'
+                f'XFProj={xf_proj},NPOP={npop},gen={gen} '
+                f'{working_dir}/Batch_Jobs/GPU_XF_Job.sh',
+                shell=True, capture_output=True) 
```

Comment:
```suggestion
    p2 = sp.run(['sbatch', f'--array=1{npop}%{batch_size} ',
                f'--job-name={run_name}',
                f'--export=ALL,WorkingDir={working_dir},RunName={run_name},'
                f'XFProj={xf_proj},NPOP={npop},gen={gen}',
                f'{working_dir}/Batch_Jobs/GPU_XF_Job.sh'],
                capture_output=True) 
```
This should be equivalent.  Avoiding `shell=True` is usually better.  I also added a comma after the `RunName=...`, since I think it was missing.

Reply:
I thought I needed `shell=True` to use `sbatch`, but it turns out I didn't.
I am now writing this part as follows:
```python
    try:
        sp.call(['sbatch', f'--array=1{npop}%{batch_size}',
                f'--job-name={run_name}',
                f'--export=ALL,WorkingDir={working_dir},RunName={run_name},'
                f'XFProj={xf_proj},NPOP={npop},gen={gen}',
                f'{working_dir}/Batch_Jobs/GPU_XF_Job.sh'],
                capture_output=True)
    except:
        print('\nError at part_b_job1.py during job submission\n')
```
because I couldn't figure out how to suppress the error message (`sbatch` not found since running locally)